applications/blackhole/src/blackhole_socket_callback.erl

```diff
@@ -91,28 +91,47 @@ handle_info(_SessionPid, _SessionId, _Msg, Context) ->
 %% @doc
 %% @end
 %%--------------------------------------------------------------------
--spec subscribe(bh_context:context(), wh_json:object()) -> cb_return().
--spec subscribe(bh_context:context(), wh_json:object(), boolean()) -> cb_return().
--spec subscribe(bh_context:context(), wh_json:object(), ne_binary(), api_binary()) -> cb_return().
-subscribe(Context, JObj) ->
-    Context1 = bh_context:subscribe(Context, JObj),
+-spec maybe_subscribe(bh_context:context(), wh_json:object()) -> cb_return().
+-spec maybe_subscribe(bh_context:context(), wh_json:object(), boolean()) -> cb_return().
+maybe_subscribe(Context, JObj) ->
+    Context1 = bh_context:from_json(Context, JObj),
     IsAuthorized = blackhole_util:is_authorized(Context1),
-    subscribe(Context1, JObj, IsAuthorized).
+    maybe_subscribe(Context1, JObj, IsAuthorized).
 
-subscribe(Context, _JObj, 'false') ->
+maybe_subscribe(Context, _JObj, 'false') ->
     {'ok', blackhole_util:respond_with_authn_failure(Context)};
-subscribe(Context, JObj, 'true') ->
-    Binding = wh_json:get_value(<<"binding">>, JObj),
-    Module = blackhole_util:get_callback_module(Binding),
-    subscribe(Context, JObj, Binding, Module).
+maybe_subscribe(Context, JObj, 'true') ->
+    Bindings = bh_context:bindings_from_json(JObj),
+    check_bindings(Context, JObj, Bindings).
 
+-spec check_bindings(bh_context:context(), wh_json:object(), ne_binaries() | ne_binary()) -> cb_return().
+check_bindings(Context, _JObj, []) ->
+    {'ok', Context};
+check_bindings(Context, JObj, [Binding|Bds]) ->
+    {'ok', Context1} = check_bindings(Context, JObj, Binding),
+    check_bindings(Context1, JObj, Bds);
+check_bindings(Context, JObj, Binding) ->
```

Comment:
Maybe a `Binding = ?NE_BINARY` will help Dialyzer; here or in `bh_context:already_bindeded/2`.